ml-proto/spec/int.ml

```diff
@@ -208,7 +208,55 @@ struct
   let ge_s x y = x >= y
   let ge_u x y = cmp_u x (>=) y
 
-  let of_string = Rep.of_string
+  (* This implementation allows leading signs and unsigned values *)
+  let of_string x =
+    let len = String.length x in
+    let power_of_two n = Rep.shift_left Rep.one n in
+    let ten = Rep.of_int 10 in
+    let fail () = failwith "int_of_string" in
+    let parse_hexdigit c =
+      int_of_char c -
+        if '0' <= c && '9' >= c then 0x30
+        else if 'a' <= c && 'f' >= c then 0x61 - 10
+        else if 'A' <= c && 'F' >= c then 0x41 - 10
+        else fail ()
+    in
+    let parse_hex offset sign =
+      let num = ref Rep.zero in
+      for i = offset to len - 1 do
+        if ge_u !num (power_of_two (Rep.bitwidth - 4)) then fail ();
+        num := Rep.logor (Rep.shift_left !num 4) (Rep.of_int (parse_hexdigit x.[i]));
+      done;
+      if sign && gt_u !num (power_of_two (Rep.bitwidth - 1)) then fail ();
+      !num
+    in
+    let parse_dec offset sign =
+      let max_upper, max_lower =
+        if sign then
+          divrem_u (power_of_two (Rep.bitwidth - 1)) ten
+        else
+          divrem_u Rep.minus_one ten
+      in
+      let num = ref Rep.zero in
+      for i = offset to len - 1 do
+        if '0' > x.[i] || '9' < x.[i] then fail ();
+        let new_digit = Rep.of_int (int_of_char x.[i] - 0x30) in
+        if gt_u !num max_upper || (!num = max_upper && gt_u new_digit max_lower) then fail ();
+        num := Rep.add (Rep.mul !num ten) new_digit
+      done;
+      !num
+    in
+    let parse_int offset sign =
+      if offset + 3 <= len && String.sub x offset 2 = "0x" then
```

Comment:
Nit: it's cheaper to just compare `x.[offset]` and x.[offset+1]` than allocating a substring.